server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -127,52 +131,238 @@ void setUpdateFrequency(TimeValue updateFrequency) {
-                    ImmutableOpenMap.Builder<String, DiskUsage> newMaxUsages = ImmutableOpenMap.builder(leastAvailableSpaceUsages);
-                    newMaxUsages.remove(removedNode.getId());
-                    leastAvailableSpaceUsages = newMaxUsages.build();
+        void execute() {
+            assert countDown.isCountedDown() == false;
+
+            logger.trace("starting async refresh");
+
+            final NodesStatsRequest nodesStatsRequest = new NodesStatsRequest("data:true");
+            nodesStatsRequest.clear();
+            nodesStatsRequest.addMetric(NodesStatsRequest.Metric.FS.metricName());
+            nodesStatsRequest.timeout(fetchTimeout);
+            client.admin().cluster().nodesStats(nodesStatsRequest, ActionListener.runAfter(new ActionListener<>() {
+                @Override
+                public void onResponse(NodesStatsResponse nodesStatsResponse) {
+                    logger.trace("received node stats response");
+
+                    for (final FailedNodeException failure : nodesStatsResponse.failures()) {
+                        final Throwable cause = failure.getCause();
+                        if (logger.isDebugEnabled()) {
+                            logger.warn(new ParameterizedMessage("failed to retrieve stats for node [{}]", failure.nodeId()), cause);
+                        } else {
+                            logger.warn("failed to retrieve stats for node [{}]: {}", failure.nodeId(), cause.getMessage());
+                        }
+                    }
+
+                    ImmutableOpenMap.Builder<String, DiskUsage> leastAvailableUsagesBuilder = ImmutableOpenMap.builder();
+                    ImmutableOpenMap.Builder<String, DiskUsage> mostAvailableUsagesBuilder = ImmutableOpenMap.builder();
+                    fillDiskUsagePerNode(adjustNodesStats(nodesStatsResponse.getNodes()),
+                            leastAvailableUsagesBuilder, mostAvailableUsagesBuilder);
+                    leastAvailableSpaceUsages = leastAvailableUsagesBuilder.build();
+                    mostAvailableSpaceUsages = mostAvailableUsagesBuilder.build();
                 }
-                if (mostAvailableSpaceUsages.containsKey(removedNode.getId())) {
-                    ImmutableOpenMap.Builder<String, DiskUsage> newMinUsages = ImmutableOpenMap.builder(mostAvailableSpaceUsages);
-                    newMinUsages.remove(removedNode.getId());
-                    mostAvailableSpaceUsages = newMinUsages.build();
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof ClusterBlockException) {
+                        logger.trace("failed to retrieve node stats", e);
+                    } else {
+                        logger.warn("failed to retrieve node stats", e);
+                    }
+                    leastAvailableSpaceUsages = ImmutableOpenMap.of();
+                    mostAvailableSpaceUsages = ImmutableOpenMap.of();
```

Comment:
I think it's bad that we simply forget everything here. Anyway, not something that needs to be addressed in this PR.

Reply:
Yeah there's still a couple of wrinkles - this is one, the fact that we don't refresh the node stats and the indices stats at the same time is another.

I think this particular bit isn't too bad: we only get here if coordinating the stats request completely fails, and I'm not sure that this can even happen. It's worse that we forget individual nodes if they time out.